src/output.c

```diff
@@ -683,11 +695,42 @@ do_output_layout_change(struct server *server)
 	}
 }
 
+void
+output_update_fallback(struct server *server)
+{
+	assert(server->fallback_output.name);
+	struct wl_list *layout_outputs = &server->output_layout->outputs;
+
+	if (wl_list_empty(layout_outputs)) {
+		wlr_log(WLR_DEBUG, "adding fallback output %s",
+			server->fallback_output.name);
+		assert(!server->fallback_output.output);
+
+		output_add_virtual(server, server->fallback_output.name,
+			&server->fallback_output.output);
+	} else if (server->fallback_output.output
+			&& wl_list_length(layout_outputs) > 1) {
+		wlr_log(WLR_DEBUG, "destroying fallback output %s",
+			server->fallback_output.name);
+
+		wlr_output_destroy(server->fallback_output.output);
+		server->fallback_output.output = NULL;
+	}
+}
+
 static void
 handle_output_layout_change(struct wl_listener *listener, void *data)
```

Comment:
Yes, the most important reason is that outputs can be enabled / disabled during runtime which removes them from the layout but does not call into `output_destroy_notify()`.

Temporarily suspending `server->new_output.link` when adding a virtual output is the only option so that we have access to the resulting `wlr_output` before enabling it. We need that for example to call `wlr_output_set_name()` which can't be called after the output has already been enabled. Maybe I should throw that part into a separate PR because it also simplifies `output_add_virtual()` logic in general (e.g. it gets rid of that temporary `server->headless.pending_output_name` char array).

That is also the reason for the new `struct wlr_output **store_wlr_output` argument, otherwise we can't identify the fallback output in `new_output_notify()` (and all handlers like `handle_output_layout_change()` that are called from there, either directly or indirectly).